### PR TITLE
Dump document vectors in IndexUtil (#318)

### PR DESCRIPTION
**Pull request by Kytabyte**
1. I changed `printDocvector` argument into `dumpDocVector` to keep the consistency.
2. This structure is the same as `DumpRawDocs`, where the input is a file with all required docids and output is a tar.gz file containing the doc vector for one document per file.
3. the usage is `sh target/appassembler/bin/IndexUtils -index /path/to/index -dumpDocVectors /path/to/docid -docVectorWeight tf.idf`, where the `-docVectorWeight` is optional and can be easily expanded to other weight method in the future.

### CODE REVIEW

File: src/main/java/io/anserini/index/IndexUtils.java [hunk 1]
```diff
@@ -105,6 +113,12 @@ public NotStoredException(String message) {
     }
   }
 
+  public class InvalidDocVectorWeightException extends Exception {
```
Comment (lintool) [inline src/main/java/io/anserini/index/IndexUtils.java]: no need for this - see below.

File: src/main/java/io/anserini/index/IndexUtils.java [hunk 2]
```diff
@@ -183,6 +197,104 @@ public void printDocumentVector(String docid) throws IOException, NotStoredExcep
...
+        outLength += (entry.getKey().getBytes(StandardCharsets.UTF_8).length
+                + entry.getValue().getBytes(StandardCharsets.UTF_8).length
+                + " \n".length());
+      }
+
+      TarArchiveEntry tarEntry = new TarArchiveEntry(new File(docid));
+      tarEntry.setSize(outLength + String.format("<DOCNO>%s</DOCNO>\n", docid).length());
+      tOut.putArchiveEntry(tarEntry);
+      tOut.write(String.format("<DOCNO>%s</DOCNO>\n", docid).getBytes());
+      for (Map.Entry<String, String> entry: docVectors.entrySet()) {
```
Comment (lintool) [inline src/main/java/io/anserini/index/IndexUtils.java]: Do you know about this?
```
Joiner.MapJoiner mapJoiner = Joiner.on(",").withKeyValueSeparator("=");
System.out.println(mapJoiner.join(map));
```
see: https://google.github.io/guava/releases/19.0/api/docs/com/google/common/base/Joiner.MapJoiner.html

**Comment by lintool** [comment]
High-level question - do we really need to compress the doc vectors?

@Peilin-Yang is that your feature request?

Seems a bit like overkill?

**Comment by Kytabyte** [comment]
NYT collection contains 1,800,000+ documents, and the dumped uncompressed TF-IDF file is around 6 GB. I think for even larger collection, like ClueWeb12, the compressed file would be useful?

**Comment by lintool** [comment]
@Kytabyte I see. Is there a use case where we would dump the entire collection? I thought we'd only do this selectively, e.g., qrels...

File: src/main/java/io/anserini/index/IndexUtils.java (hunk 2 above)
Comment (Peilin-Yang) [reply src/main/java/io/anserini/index/IndexUtils.java]: How about 
```
String s = ranking.entrySet().stream().map(e -> e.getKey()+" "+e.getValue()).collect(joining("\n"));
```
?
Guava is less and less popular after Java 8 and Scala....

Comment (lintool) [reply src/main/java/io/anserini/index/IndexUtils.java]: Oh really? Should we kill it as a dependency then?

It was very much in use when I was at Twitter. Every method started with Guava precondition checks...

**Comment by Peilin-Yang** [comment]
For raw documents we can benefit from tar-balling the outputs.
I am not sure about the exact use case here but enabling compression makes it more scalable.

File: src/main/java/io/anserini/index/IndexUtils.java (hunk 2 above)
Comment (Peilin-Yang) [reply src/main/java/io/anserini/index/IndexUtils.java]: I think we could file an issue for it.
In Java 8 and Scala/Rust, the introduction of `Option` forces the programmers to deal with `Null` which eliminates most of the precondition checks

File: src/main/java/io/anserini/index/IndexUtils.java [hunk 3]
```diff
@@ -183,6 +194,98 @@ public void printDocumentVector(String docid) throws IOException, NotStoredExcep
...
+
+    InputStream in = getReadFileStream(reqDocidsPath);
+    BufferedReader bRdr = new BufferedReader(new InputStreamReader(in));
+    FileOutputStream fOut = new FileOutputStream(new File(outFileName));
+    BufferedOutputStream bOut = new BufferedOutputStream(fOut);
+    GzipCompressorOutputStream gzOut = new GzipCompressorOutputStream(bOut);
+    TarArchiveOutputStream tOut = new TarArchiveOutputStream(gzOut);
+
+    Map<Term, Integer> docFreqMap = new HashMap<>();
+    Map<String, String> docVectors;
```
Comment (Peilin-Yang) [inline src/main/java/io/anserini/index/IndexUtils.java]: The definition can be moved to inside the loop?
We will introduce GC cycles anyway...

**Comment by Kytabyte** [comment]
> Could you please also fix #325 in this PR?
And also make the loggings consistent for dumpDocVectors and dumpRawDocs?

will do!

File: src/main/java/io/anserini/index/IndexUtils.java (hunk 3 above)
Comment (Kytabyte) [reply src/main/java/io/anserini/index/IndexUtils.java]: The `docFreqMap` is a global variable for the entire loops. It stores the document frequency of every `Term` when it occurs and we query it from indexing, thus we can read it from memory next time we see this term. It'll be much slower if we query DF every time...

Comment (Peilin-Yang) [reply src/main/java/io/anserini/index/IndexUtils.java]: Yes, what I meant is `docVectors`.

Comment (Kytabyte) [reply src/main/java/io/anserini/index/IndexUtils.java]: Oh, sure, will do.

**Comment by Peilin-Yang** [comment]
@lintool do you have any further comments?

**Pull request merged**